Index: linux-3.18.14/drivers/mtd/nand/ar934x_nfc.c
===================================================================
--- linux-3.18.14.orig/drivers/mtd/nand/ar934x_nfc.c
+++ linux-3.18.14/drivers/mtd/nand/ar934x_nfc.c
@@ -1437,6 +1437,10 @@ ar934x_nfc_probe(struct platform_device
 		nand->ecc.mode = NAND_ECC_SOFT;
 		break;

+	case AR934X_NFC_ECC_SOFT_BCH:
+		nand->ecc.mode = NAND_ECC_SOFT_BCH;
+		break;
+
 	case AR934X_NFC_ECC_HW:
 		ret = ar934x_nfc_setup_hwecc(nfc);
 		if (ret)
Index: linux-3.18.14/include/linux/platform/ar934x_nfc.h
===================================================================
--- linux-3.18.14.orig/include/linux/platform/ar934x_nfc.h
+++ linux-3.18.14/include/linux/platform/ar934x_nfc.h
@@ -20,6 +20,7 @@ struct mtd_partition;
 enum ar934x_nfc_ecc_mode {
 	AR934X_NFC_ECC_SOFT = 0,
 	AR934X_NFC_ECC_HW,
+	AR934X_NFC_ECC_SOFT_BCH,
 };

 struct ar934x_nfc_platform_data {
